mautic/mautic
# API for webhooks

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/190
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation/pull/61
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/1591, https://github.com/mautic/mautic/issues/4151
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR implements the REST Hooks based on the [Zapier docs](https://zapier.com/developer/documentation/v2/rest-hooks/)

Todo:
- [x] Add WebhookBundle to PHPSTAN CI
- [x] Create CRUD API endpoints
- [x] Update API library (https://github.com/mautic/api-library/pull/112)
- [x] Improve logs with error messages and runtime
- [x] Allow to configure the CURL timeout
- [x] Stop the webhook after X bad responses
- [x] Respect 410 responses on payload delivery and remove subscriptions
- [x] Unit tests (at least one)
- [x] Update dev docs
- [x] Update test payloads (https://github.com/mautic/mautic/issues/4151)
- [x] Update user docs

#### Steps to test this PR:
1. Apply this PR
2. Clear cache
3. Run migrations
4. Apply https://github.com/mautic/api-library/pull/112
5. Run `phpunit --filter WebhooksTest` in the API library root dir to test the new API endpoint.
6. Go to the webhook UI and test the new webhook logs. For example with Request Bin. You should see how long the request took and if it will take more than 15 seconds, the status will be `timeout`. This can be configured in the config.php if you want to test that with a different value.
7. Test the webhooks also in the queue mode (switch in the configuration), make the action from your webhook and run command `app/console mautic:webhooks:process`
8. If you know how, try to return other than 2** HTTP response for 100 times or modify `webhook_disable_limit` config option to some smaller number. After that limit the webhook should unpublish itself and your user should get a notification.
9. If the HTTP response status is 410, the webhook unpublishes itself and again notifies the user. Test it if you know how.
